File: src/Sample/Plumix.Sample/Demos/General/BlocCounterDemoPage.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Avalonia;
using Avalonia.Media;
using Plumix.Bloc;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): pub.dev/packages/bloc; pub.dev/packages/flutter_bloc (sample parity demo)

namespace Plumix;

public sealed class BlocCounterDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new BlocCounterDemoPageState();
    }
}

internal sealed class BlocCounterDemoPageState : State
{
    private string _milestoneMessage = "Milestone listener: waiting for count divisible by 5.";

    public override Widget Build(BuildContext context)
    {
        return new BlocProvider<BlocCounterBloc>(
            create: static _ => new BlocCounterBloc(),
            child: new BlocListener<BlocCounterBloc, BlocCounterState>(
                listenWhen: static (previous, next) =>
                    previous.Count != next.Count
                    && next.Count != 0
                    && next.Count % 5 == 0,
                listener: (_, state) =>
                {
                    SetState(() =>
                    {
                        _milestoneMessage = $"Milestone listener: count={state.Count} at {DateTime.Now:HH:mm:ss}.";
                    });
                },
                child: new Column(
                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                    spacing: 10,
                    children:
                    [
                        new Text("Bloc counter demo", fontSize: 20, color: Colors.Black),
                        new Text(
                            "BlocProvider + BlocBuilder + BlocListener + BlocSelector. Refresh event uses restartable transformer.",
                            fontSize: 14,
                            color: Color.Parse("#8A000000")),
                        new Text(_milestoneMessage, fontSize: 12, color: Color.Parse("#FF607D8B")),
                        new BlocSelector<BlocCounterBloc, BlocCounterState, int>(
                            selector: static state => state.Count,
                            builder: static (_, count) =>
                                new Text($"count={count}", fontSize: 18, color: Colors.DarkSlateBlue)),
                        new BlocBuilder<BlocCounterBloc, BlocCounterState>(
                            buildWhen: static (previous, next) => previous.IsLoading != next.IsLoading,
                            builder: static (_, state) => new Text(
                                state.IsLoading
                                    ? "loading=true (refresh in-flight, restartable)"
                                    : "loading=false",
                                fontSize: 12,
                                color: state.IsLoading ? Color.Parse("#FF8E24AA") : Color.Parse("#FF2E7D32"))),
                        new BlocCounterActionButtons(),
                    ])));
    }

    internal static Widget BuildActionButton(string label, Color background, Action onPressed)
    {
        return new Expanded(
            child: new TextButton(
                onPressed: onPressed,
                backgroundColor: background,
                foregroundColor: Colors.Black,
                minHeight: 38,
                padding: new Thickness(10, 8),
                borderRadius: BorderRadius.Circular(8),
                child: new Text(label, fontSize: 12, textAlign: TextAlign.Center)));
    }
}

internal sealed class BlocCounterActionButtons : StatelessWidget
{
    public override Widget Build(BuildContext context)
    {
        var bloc = context.Read<BlocCounterBloc>();

        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new Row(
                    spacing: 8,
                    children:
                    [
                        BlocCounterDemoPageState.BuildActionButton(
                            label: "-1",
                            background: Color.Parse("#FFECEFF1"),
                            onPressed: () => bloc.Add(new DecrementCounterEvent())),
                        BlocCounterDemoPageState.BuildActionButton(
                            label: "+1",
                            background: Color.Parse("#FFE3F2FD"),
                            onPressed: () => bloc.Add(new IncrementCounterEvent())),
                        BlocCounterDemoPageState.BuildActionButton(
                            label: "+5",
                            background: Color.Parse("#FFE8F5E9"),
                            onPressed: () => bloc.Add(new IncrementByCounterEvent(5))),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BlocCounterDemoPageState.BuildActionButton(
                            label: "Refresh +10 (350ms)",
                            background: Color.Parse("#FFFFF3E0"),
                            onPressed: () => bloc.Add(new RefreshCounterEvent(10, 350))),
                        BlocCounterDemoPageState.BuildActionButton(
                            label: "Refresh +10 (80ms)",
                            background: Color.Parse("#FFF3E5F5"),
                            onPressed: () => bloc.Add(new RefreshCounterEvent(10, 80))),
                    ]),
                BlocCounterDemoPageState.BuildActionButton(
                    label: "Reset",
                    background: Color.Parse("#FFFFEBEE"),
                    onPressed: () => bloc.Add(new ResetCounterEvent())),
            ]);
    }
}

internal readonly record struct BlocCounterState(int Count, bool IsLoading);

internal abstract record CounterEvent;
internal sealed record IncrementCounterEvent : CounterEvent;
internal sealed record DecrementCounterEvent : CounterEvent;
internal sealed record IncrementByCounterEvent(int Delta) : CounterEvent;
internal sealed record RefreshCounterEvent(int Delta, int DelayMs) : CounterEvent;
internal sealed record ResetCounterEvent : CounterEvent;

internal sealed class BlocCounterBloc : Bloc<CounterEvent, BlocCounterState>
{
    public BlocCounterBloc() : base(new BlocCounterState(0, IsLoading: false))
    {
        On<IncrementCounterEvent>(HandleIncrement);
        On<DecrementCounterEvent>(HandleDecrement);
        On<IncrementByCounterEvent>(HandleIncrementBy);
        On<ResetCounterEvent>(HandleReset);
        On<RefreshCounterEvent>(HandleRefreshAsync, EventTransformers.Restartable<RefreshCounterEvent>());
    }

    private ValueTask HandleIncrement(IncrementCounterEvent @event, IEmitter<BlocCounterState> emitter, CancellationToken cancellationToken)
    {
        emitter.Emit(State with { Count = State.Count + 1 });
        return ValueTask.CompletedTask;
    }

    private ValueTask HandleDecrement(DecrementCounterEvent @event, IEmitter<BlocCounterState> emitter, CancellationToken cancellationToken)
    {
        emitter.Emit(State with { Count = State.Count - 1 });
        return ValueTask.CompletedTask;
    }

    private ValueTask HandleIncrementBy(IncrementByCounterEvent @event, IEmitter<BlocCounterState> emitter, CancellationToken cancellationToken)
    {
        emitter.Emit(State with { Count = State.Count + @event.Delta });
        return ValueTask.CompletedTask;
    }

    private ValueTask HandleReset(ResetCounterEvent @event, IEmitter<BlocCounterState> emitter, CancellationToken cancellationToken)
    {
        emitter.Emit(new BlocCounterState(0, IsLoading: false));
        return ValueTask.CompletedTask;
    }

    private async ValueTask HandleRefreshAsync(
        RefreshCounterEvent @event,
        IEmitter<BlocCounterState> emitter,
        CancellationToken cancellationToken)
    {
        emitter.Emit(State with { IsLoading = true });
        await Task.Delay(@event.DelayMs, cancellationToken);
        emitter.Emit(new BlocCounterState(State.Count + @event.Delta, IsLoading: false));
    }
}

File: dart_sample/lib/demos/general/bloc_counter_demo_page.dart
import 'package:bloc_concurrency/bloc_concurrency.dart';
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

class BlocCounterDemoPage extends StatefulWidget {
  const BlocCounterDemoPage({super.key});

  @override
  State<BlocCounterDemoPage> createState() => _BlocCounterDemoPageState();

  static Widget buildActionButton({
    required String label,
    required Color background,
    required VoidCallback onPressed,
  }) {
    return Expanded(
      child: TextButton(
        onPressed: onPressed,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 38),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

class _BlocCounterDemoPageState extends State<BlocCounterDemoPage> {
  String _milestoneMessage =
      'Milestone listener: waiting for count divisible by 5.';

  @override
  Widget build(BuildContext context) {
    return BlocProvider<BlocCounterBloc>(
      create: (_) => BlocCounterBloc(),
      child: BlocListener<BlocCounterBloc, BlocCounterState>(
        listenWhen: (BlocCounterState previous, BlocCounterState next) =>
            previous.count != next.count &&
            next.count != 0 &&
            next.count % 5 == 0,
        listener: (_, BlocCounterState state) {
          setState(() {
            _milestoneMessage =
                'Milestone listener: count=${state.count} at ${_formatNow()}.';
          });
        },
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 10,
          children: <Widget>[
            const Text(
              'Bloc counter demo',
              style: TextStyle(fontSize: 20, color: Colors.black),
            ),
            const Text(
              'BlocProvider + BlocBuilder + BlocListener + BlocSelector. Refresh event uses restartable transformer.',
              style: TextStyle(fontSize: 14, color: Colors.black54),
            ),
            Text(
              _milestoneMessage,
              style: const TextStyle(fontSize: 12, color: Color(0xFF607D8B)),
            ),
            BlocSelector<BlocCounterBloc, BlocCounterState, int>(
              selector: (BlocCounterState state) => state.count,
              builder: (_, int count) => Text(
                'count=$count',
                style: const TextStyle(fontSize: 18, color: Colors.blueGrey),
              ),
            ),
            BlocBuilder<BlocCounterBloc, BlocCounterState>(
              buildWhen: (BlocCounterState previous, BlocCounterState next) =>
                  previous.isLoading != next.isLoading,
              builder: (_, BlocCounterState state) => Text(
                state.isLoading
                    ? 'loading=true (refresh in-flight, restartable)'
                    : 'loading=false',
                style: TextStyle(
                  fontSize: 12,
                  color: state.isLoading
                      ? const Color(0xFF8E24AA)
                      : const Color(0xFF2E7D32),
                ),
              ),
            ),
            const _BlocCounterActionButtons(),
          ],
        ),
      ),
    );
  }

  String _formatNow() {
    final DateTime now = DateTime.now();
    return '${_twoDigits(now.hour)}:${_twoDigits(now.minute)}:${_twoDigits(now.second)}';
  }

  String _twoDigits(int value) => value.toString().padLeft(2, '0');
}

class _BlocCounterActionButtons extends StatelessWidget {
  const _BlocCounterActionButtons();

  @override
  Widget build(BuildContext context) {
    final BlocCounterBloc bloc = context.read<BlocCounterBloc>();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        Row(
          spacing: 8,
          children: <Widget>[
            BlocCounterDemoPage.buildActionButton(
              label: '-1',
              background: const Color(0xFFECEFF1),
              onPressed: () => bloc.add(const DecrementCounterEvent()),
            ),
            BlocCounterDemoPage.buildActionButton(
              label: '+1',
              background: const Color(0xFFE3F2FD),
              onPressed: () => bloc.add(const IncrementCounterEvent()),
            ),
            BlocCounterDemoPage.buildActionButton(
              label: '+5',
              background: const Color(0xFFE8F5E9),
              onPressed: () => bloc.add(const IncrementByCounterEvent(5)),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            BlocCounterDemoPage.buildActionButton(
              label: 'Refresh +10 (350ms)',
              background: const Color(0xFFFFF3E0),
              onPressed: () => bloc.add(const RefreshCounterEvent(10, 350)),
            ),
            BlocCounterDemoPage.buildActionButton(
              label: 'Refresh +10 (80ms)',
              background: const Color(0xFFF3E5F5),
              onPressed: () => bloc.add(const RefreshCounterEvent(10, 80)),
            ),
          ],
        ),
        BlocCounterDemoPage.buildActionButton(
          label: 'Reset',
          background: const Color(0xFFFFEBEE),
          onPressed: () => bloc.add(const ResetCounterEvent()),
        ),
      ],
    );
  }
}

class BlocCounterState {
  const BlocCounterState({required this.count, required this.isLoading});

  final int count;
  final bool isLoading;

  BlocCounterState copyWith({int? count, bool? isLoading}) {
    return BlocCounterState(
      count: count ?? this.count,
      isLoading: isLoading ?? this.isLoading,
    );
  }
}

sealed class CounterEvent {
  const CounterEvent();
}

final class IncrementCounterEvent extends CounterEvent {
  const IncrementCounterEvent();
}

final class DecrementCounterEvent extends CounterEvent {
  const DecrementCounterEvent();
}

final class IncrementByCounterEvent extends CounterEvent {
  const IncrementByCounterEvent(this.delta);

  final int delta;
}

final class RefreshCounterEvent extends CounterEvent {
  const RefreshCounterEvent(this.delta, this.delayMs);

  final int delta;
  final int delayMs;
}

final class ResetCounterEvent extends CounterEvent {
  const ResetCounterEvent();
}

class BlocCounterBloc extends Bloc<CounterEvent, BlocCounterState> {
  BlocCounterBloc()
    : super(const BlocCounterState(count: 0, isLoading: false)) {
    on<IncrementCounterEvent>(_onIncrement);
    on<DecrementCounterEvent>(_onDecrement);
    on<IncrementByCounterEvent>(_onIncrementBy);
    on<ResetCounterEvent>(_onReset);
    on<RefreshCounterEvent>(_onRefresh, transformer: restartable());
  }

  void _onIncrement(
    IncrementCounterEvent event,
    Emitter<BlocCounterState> emit,
  ) {
    emit(state.copyWith(count: state.count + 1));
  }

  void _onDecrement(
    DecrementCounterEvent event,
    Emitter<BlocCounterState> emit,
  ) {
    emit(state.copyWith(count: state.count - 1));
  }

  void _onIncrementBy(
    IncrementByCounterEvent event,
    Emitter<BlocCounterState> emit,
  ) {
    emit(state.copyWith(count: state.count + event.delta));
  }

  void _onReset(ResetCounterEvent event, Emitter<BlocCounterState> emit) {
    emit(const BlocCounterState(count: 0, isLoading: false));
  }

  Future<void> _onRefresh(
    RefreshCounterEvent event,
    Emitter<BlocCounterState> emit,
  ) async {
    emit(state.copyWith(isLoading: true));
    await Future<void>.delayed(Duration(milliseconds: event.delayMs));
    if (emit.isDone) {
      return;
    }
    emit(BlocCounterState(count: state.count + event.delta, isLoading: false));
  }
}
